### ISSUE TITLE
Max-num-batched-tokens as a second memory-wall knob: a measured 0.09 -> 3.50 GiB headroom delta at 1536 chunk (and why MemAvailable thresholds do not port between fleets)

### ISSUE THREAD

**Issue by ZackO2o** [MiaAI-Lab/DeepSeek-v4.1-Flash-EXL3-2x-DGX-Sparks#19]
**Scope note, added after posting:** most of what I first wrote here overlaps existing threads —
the missing pinned `.so` is [#17](https://github.com/MiaAI-Lab/DeepSeek-v4.1-Flash-EXL3-2x-DGX-Sparks/issues/17)
(where I have since posted our independent confirmation of the self-build route), and independent 2-node
results are already in [#9](https://github.com/MiaAI-Lab/DeepSeek-v4.1-Flash-EXL3-2x-DGX-Sparks/issues/9).
Rather than duplicate those, this issue is narrowed to the two things neither thread covers, both of
which came out of chasing a boot failure on our pair.

## 1. `--max-num-batched-tokens` moved our memory wall further than gmu did — with a magnitude

The README already states the effect qualitatively:

> *"At the shipped 1536-token chunk with images on, head low-water is higher than the 2048-chunk run
> above — a larger chunk finishes the prefill in fewer scheduler passes, which more than pays for the
> bigger per-chunk activation."*

We hit this from the failure side and can attach a number to it:

| `MAX_NUM_BATCHED_TOKENS` | head `MemAvailable` at boot |
|---|---|
| 1024 | **0.09 GiB** |
| **1536** | **3.50 GiB** |

The sequence matters, because the first two attempts pointed the wrong way. At gmu 0.85 our preflight
came up short by ~1 GiB. Stepping gmu *down* (0.88 -> 0.86 -> 0.84) **shifted** the wall rather than
removing it: each step bought back roughly what the smaller pool gave up, and every restart cost
8-12 minutes. What actually restored headroom was raising the chunk to 1536, after which gmu 0.85
booted comfortably. Since 1536 is also the value the vision tower needs, on this model family it is
not a trade to weigh.

Worth noting against [#9](https://github.com/MiaAI-Lab/DeepSeek-V4.1-Flash-EXL3-2x-DGX-Sparks/issues/9),
which ran `MAX_NUM_BATCHED_TOKENS=1024`: if that configuration was also tight on headroom, the chunk
size may be why.

**Suggestion:** attach the number to the README sentence. "A larger chunk gives more headroom" reads
backwards enough that most operators will not act on it without one — we did not, which is why we spent
several restart cycles walking gmu in the wrong direction first.

## 2. `MemAvailable` is a floor, not a gauge, and its threshold does not port between fleets

On a 4-node fleet we measured **10.3 GiB** `MemAvailable` after a full 1M request. On this 2-node pair
the same quantity reads **1.1-2.4 GiB** while serving normally, and frequently ~0, with **zero** real
OOM kills (`journalctl -k` free of `oom-kill` / `Out of memory: Killed`).

The mechanism is the one the README already gives for the pool behaving oddly — the weights live in
driver allocations not charged to any process RSS — so the kernel's "available" estimate has little
left to count. Two practical consequences:

- a watchdog keyed on `MemAvailable` needs a **per-fleet** threshold; one ported between these two
  fleets would either never fire or fire constantly;
- the signals that *do* survive the port are a real `oom-kill` line from the kernel, a container
  health/restart transition, and the **post-long-prefill** low-water mark (the floor is set by a long
  prefill, not by boot).

For reference, the README's own figures — 4.07-4.21 GiB warmed, 2.1 GiB at 601k — and ours (2.35 GiB
after a long request) are the same order of magnitude, so this is not a claim that our pair is
tighter; it is a claim that the metric does not carry a usable absolute threshold.

## Reference: what this pair measured, for anyone comparing

Config: `MAX_MODEL_LEN=600000`, `MAX_NUM_SEQS=2`, `MAX_NUM_BATCHED_TOKENS=1536`, `GPU_MEM_UTIL=0.85`,
DSpark k=3, cooperative MoE from a local gated build, Engram packed to node-local NVMe, abliterated
overlay applied with `ENGRAM_DIR` left native. Base is EXL3 2.9 bpw.

Prefill 966-1021 tok/s (25K and 100K prompts, packed Engram, zero cached tokens) — consistent with the
README's 979-1029. Decode is strongly load-dependent here: 31.1 tok/s on a counting sequence, 36.0 on
prose, 49.8 writing code, all temperature 0 reading `usage.completion_tokens`. We are not claiming a
single headline decode number for that reason; the ~1.6x spread is why the README separating Poetry
from Coding is the right shape for that table.

**Comment by chortya** [comment]
Adding a data point on the 1536 chunk at the extreme end, since the README flags the shipped 1536/96 default as not yet revalidated at 600k.

Exact-size synthetic prompt, **599,840 input tokens** (600,000 − 160 output), single request, thinking off, three-key recall check. Config: shipped defaults (`MAX_NUM_BATCHED_TOKENS=1536`, `DSV41_IO_THREADS=96`, `MAX_MODEL_LEN=600000`, `KV_CACHE_MEMORY_BYTES=2684354560`, `gpu-memory-utilization 0.85`, `PREFIX_CACHE_RETENTION_INTERVAL=4096`), plus a local opt-in GB10 decode-MoE overlay.

| Run | TTFT | Prefill tok/s | Recall | Cached tokens |
|---|---|---|---|---|
| cold | 747.8 s | **802.1** | correct (3/3) | 0 |
| warm | 897.8 s | 668.1 | correct (3/3) | 0 |

Head `MemAvailable` stayed ~4.6 GiB through the prefill — well above the ~2.1 GiB floor the README documents at 601k. No OOM, no allocator failure, no engine death. So on this pair the 1536/96 default **does** survive a full 600k single prefill.

Two things worth flagging for the threshold discussion:

1. **Near-limit prompts do not cache.** The warm run had zero cache hits. A 599,840-token prompt consumes ~76% of the 785,676-token KV pool (`Maximum concurrency for 600,000 tokens per request: 1.31x` at boot), so the prefix is evicted between runs. That is capacity, not a bug — but it means `PREFIX_CACHE_RETENTION_INTERVAL=4096` buys nothing at the very top of the context window.
2. **This is single-request only.** Two concurrent 600k prefills is the tightest memory case and is untested here; that is where I would expect the 1536 chunk to be closest to the wall.

On your point that MemAvailable thresholds do not port between fleets: agreed, and this pair is a concrete example — we sit at 4.6 GiB where another fleet might sit at 0.09 GiB at the same chunk size. The useful invariant is the *delta* a chunk-size change produces, not the absolute floor.

Full field notes: deployment report on the pair (happy to paste the exact `.env`).

**Comment by Capicua25x** [comment]
Adding a data point at the top of the window, from the **no-display-unlock** side (the #23 route gets there with the 2 GB display-RAM unlock instead — this one stays on ordinary memory).

**Config** (recipe `6f7d159`, driver 595.84, no display region registered):

| Knob | Value |
|---|---|
| `MAX_MODEL_LEN` | `1048576` |
| `KV_CACHE_MEMORY_BYTES` | `3800000000` → boot reports **1,105,949-token pool** (1.05× at 1M) |
| `MAX_NUM_BATCHED_TOKENS` / `LONG_PREFILL_TOKEN_THRESHOLD` | `1024` / `768` |
| `MAX_NUM_SEQS` | `6` |

**Results** (single request, thinking off):

- **1.03M needle passes**: 1,030,036 prompt tokens, three-key recall correct, TTFT 1516.7 s, prefill **679 tok/s**.
- **The chunk size is what moves the wall.** Same pool with `MAX_NUM_BATCHED_TOKENS=3072`: the prefill **wedges at ~140k tokens** — no error, engine stays healthy, client abort clears it (no restart). At `1024` the same 1.03M prefill completes. Consistent with this thread and chortya's 1536/600k run above: what matters is headroom at the *end* of a long prefill; on our pair the 3072 chunk exhausted it.
- **The cost is prefill-only**: cold ~840 tok/s (2k–64k) vs ~1030 at 3072. Decode is unaffected (single-stream median on this config: 42.4 tok/s).
- **Concurrency follows the pool, not `MAX_MODEL_LEN`**: c1 at a full 1M; c2 while the sum of live contexts ≤ the pool. To answer chortya's open case: two concurrent 600k prefills need ≥1.2M tokens of pool → ~4.2 GB by the same ratio (untested — we stopped at 1M/c1).

Worth cross-referencing: the sibling DCP2/3bpw recipe bounds this transient directly with `VLLM_SPARSE_INDEXER_MAX_LOGITS_MB=128` rather than shrinking the chunk (it treats bounded 128 MiB indexer logits as a requirement for long contexts). I haven't tested whether this recipe's build honors that var — if it does, it's a cleaner fix than a small chunk.

So 1M per request doesn't need the display reservation — it needs the pool to cover one full request plus whatever stays resident, and a chunk small enough that the indexer transient doesn't eat the head at long prefixes.

For completeness, the `.env` lines behind that run:

```bash
MAX_MODEL_LEN=1048576
KV_CACHE_MEMORY_BYTES=3800000000
MAX_NUM_BATCHED_TOKENS=1024
LONG_PREFILL_TOKEN_THRESHOLD=768
MAX_NUM_SEQS=6
```
